Display Order Lines on Chart

ImpossiblePairs opened the pull request:
Display of lines for each order in the current market for a connected account, as well as the ability to cancel the order, and the ability to modify the order price by dragging the line and confirming.

Added a modify order function to allow change orders to be completed in one transaction.

Demo below:

https://user-images.githubusercontent.com/47860274/127901490-b2911239-bb54-42a7-8418-147d994de2e0.mp4

Please let me know any feedback.

ImpossiblePairs committed:
Initial Ideas

Basic order line options to display added.

Button functionality not implemented.

ImpossiblePairs committed:
Working Prototype

Orders for market now load onto the chart, and can be set, changed, and deleted.

Still several UI items to consider including:
- Removal of orders when switching accounts/disconnecting (current method to do this requires the chart to be reloaded)
- Handling of IoC/Market buys (intention will be to only show limit orders as this makes sense)

ImpossiblePairs committed:
TV Order Lines & Modify Transaction

Updated orderlines to be persistent, but purge when changing account, market, or disconnecting/reconnecting.

Added ability to modify a transaction (cancel and place an order in it's place) so only a single transaction is required. This should extend functionality outside of just the charting tool.

ImpossiblePairs committed:
Merge branch 'blockworks-foundation:main' into tradingview-order-lines

ImpossiblePairs committed:
Merge branch 'tradingview-order-lines' of https://github.com/ImpossiblePairs/mango-ui-v2 into tradingview-order-lines

ImpossiblePairs committed:
Merge branch 'blockworks-foundation:main' into tradingview-order-lines

ImpossiblePairs committed:
Update useEffect orderline refreshing

Removed ordersLength useState as it appeared to be triggering a reset of the orderlines halfway through a transaction and was redundant as selectedMarginAccount carries the data necessary.

mschneider commented on components/TradingView/index.tsx:
```diff
@@ -29,11 +32,21 @@ export interface ChartContainerProps {
 
-// export interface ChartContainerState {}
+const lines = new Map()
+let markPrice = 0
```
I don't see a reason why this shouldn't be solved with the useState hook. Or am I missing something?

mschneider commented on components/TradingView/index.tsx:
```diff
@@ -123,29 +196,125 @@ const TVChartContainer = () => {
+      .setQuantityTextColor('#F2C94C')
+      .setCancelButtonBorderColor(order.side == 'buy' ? '#AFD803' : '#E54033')
+      .setCancelButtonBackgroundColor('#000000')
+      .setCancelButtonIconColor('#F2C94C')
```
@saml33 can you propose how we can integrate this with the themes?

saml33 replied:
Yep, I think we can do it with this https://tailwindcss.com/docs/configuration#referencing-in-java-script

ImpossiblePairs committed:
Replaced Local Variables with State

Removed the lines and markPrice local variables and replaced with useState objects instead.
